hooks/_common.sh

```diff
@@ -233,38 +244,57 @@ function common::per_dir_hook {
           change_dir_in_unique_part="delegate_chdir"
         fi
         ;;
+      --parallelism-limit)
+        # this flag will limit the number of parallel processes
+        parallelism_limit=$((value))
+        ;;
     esac
   done
 
-  # preserve errexit status
-  shopt -qo errexit && ERREXIT_IS_SET=true
-  # allow hook to continue if exit_code is greater than 0
-  set +e
+  if [[ ! $parallelism_limit ]]; then
+    parallelism_limit=$((CPU - 1))
+  elif [[ $parallelism_limit -le 1 ]]; then
+    parallelism_limit=1
+    parallelism_disabled=true
+  fi
+
   local final_exit_code=0
+  local pids=()
+
+  mapfile -t dir_paths_unique < <(echo "${dir_paths[@]}" | tr ' ' '\n' | sort -u)
+  local length=${#dir_paths_unique[@]}
+  local last_index=$((${#dir_paths_unique[@]} - 1))
+  # run hook for each path in parallel
+  for ((i = 0; i < length; i++)); do
+    dir_path="${dir_paths_unique[$i]//__REPLACED__SPACE__/ }"
+    {
+      if [[ $change_dir_in_unique_part == false ]]; then
+        pushd "$dir_path" > /dev/null
+      fi
 
-  # run hook for each path
-  for dir_path in $(echo "${dir_paths[*]}" | tr ' ' '\n' | sort -u); do
-    dir_path="${dir_path//__REPLACED__SPACE__/ }"
+      per_dir_hook_unique_part "$dir_path" "$change_dir_in_unique_part" "$parallelism_disabled" "${args[@]}"
+    } &
+    pids+=("$!")
 
-    if [[ $change_dir_in_unique_part == false ]]; then
-      pushd "$dir_path" > /dev/null || continue
-    fi
+    if [[ $parallelism_disabled == true ]] ||
+      [ "$i" -ne 0 ] && [ $((i % parallelism_limit)) -eq 0 ] || # don't stop on first iteration when parallelism_limit>1
+      [ "$i" -eq $last_index ]; then
```

Comment:
I'm not entirely sure about the details of conditional, but I feel like conditional expressions on line 280 (`[ "$i" -ne 0 ] && [ $((i % parallelism_limit)) -eq 0 ]`) should be grouped (and if yes, then a change to third conditional (`[` -> `[[`) for consistency with previous two; also removing double quotes as they are not needed inside `[[ ... ]]`) 🤔 
```suggestion
    if [[ $parallelism_disabled == true ]] ||
      [[ $i -ne 0 && $((i % parallelism_limit)) -eq 0 ]] || # don't stop on first iteration when parallelism_limit>1
      [[ $i -eq $last_index ]]; then
```